smarts/core/utils/logging.py

```diff
@@ -162,3 +166,51 @@ def suppress_websocket():
         _logger.addFilter(websocket_filter)
         yield
         _logger.removeFilter(websocket_filter)
+
+
+def diff_unpackable(obj, other_obj):
+    """Do an asserted comparision of an object that is able to be unpacked. This works with nested collections:
+        dictionaries, namedtuples, tuples, lists, numpy arrays, and dataclasses.
+    Raises:
+        AssertionError: if objects do not match.
+    """
+    obj_unpacked = unpack(obj)
+    other_obj_unpacked = unpack(other_obj)
+
+    def sort(orig_value):
+        value = orig_value
+        if isinstance(value, (dict, defaultdict)):
+            return dict(sorted(value.items(), key=lambda item: item[0]))
+        try:
+            s = sorted(value, key=lambda item: item[0])
+        except IndexError:
+            s = sorted(value)
+        except (KeyError, TypeError):
+            s = value
+        return s
+
+    def process(o, oo, o_oo):
```

Comment:
Perhaps we could use more descriptive names here instead of `o`, `oo`, and `o_oo`.

Reply:
Done.